lib/streamlit/scriptrunner/script_requests.py

```diff
@@ -0,0 +1,177 @@
+            raise RuntimeError("RerunData is only set for RERUN requests.")
+        return cast(RerunData, self._rerun_data)
+
+
+class ScriptRequests:
+    """An interface for communicating with a ScriptRunner. Thread-safe.
+
+    AppSession makes requests of a ScriptRunner through this class, and
+    ScriptRunner handles those requests.
+    """
+
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._state = ScriptRequestType.CONTINUE
+        self._rerun_data = RerunData()
+
+    def request_stop(self) -> None:
+        """Request that the ScriptRunner stop running. A stopped ScriptRunner
+        can't be used anymore. STOP requests succeed unconditionally.
+        """
+        with self._lock:
+            self._state = ScriptRequestType.STOP
+
+    def request_rerun(self, new_data: RerunData) -> bool:
+        """Request that the ScriptRunner rerun its script.
+
+        If the ScriptRunner has been stopped, this request can't be honored:
+        return False.
+
+        Otherwise, record the request and return True. The ScriptRunner will
+        handle the rerun request as soon as it reaches an interrupt point.
+        """
+
+        with self._lock:
+            if self._state == ScriptRequestType.STOP:
+                # We can't rerun after being stopped.
+                return False
+
+            if self._state == ScriptRequestType.CONTINUE:
+                # If we're running, we can handle a rerun request
+                # unconditionally.
+                self._state = ScriptRequestType.RERUN
+                self._rerun_data = new_data
+                return True
+
+            if self._state == ScriptRequestType.RERUN:
+                # If we have an existing Rerun request, we coalesce this
+                # new request into it.
+                if self._rerun_data.widget_states is None:
+                    # The existing request's widget_states is None, which
+                    # means it wants to rerun with whatever the most
+                    # recent script execution's widget state was.
+                    # We have no meaningful state to merge with, and
+                    # so we simply overwrite the existing request.
+                    self._rerun_data = new_data
+                    return True
+
+                if new_data.widget_states is not None:
+                    # Both the existing and the new request have
+                    # non-null widget_states. Merge them together.
+                    coalesced_states = coalesce_widget_states(
+                        self._rerun_data.widget_states, new_data.widget_states
+                    )
+                    self._rerun_data = RerunData(
+                        query_string=new_data.query_string,
+                        widget_states=coalesced_states,
+                    )
+                    return True
+
+                # `old_data.widget_states is not None and data.widget_states is None` -
```

Comment:
Since this is intended as a reminder of the values of these things at this point in the code, rather than being code itself, I think it would be better as a brief english sentence rather than python